src/cider/nrepl/middleware/debug.clj

```diff
@@ -1,69 +1,197 @@
+;;;   - the keyword :expression, in which case a single sexp must be
+;;;     returned (as a string).
+
+;;;; ## Internal breakpoint logic
+;;; Variables and functions used for navigating between breakpoints.
+(def ^:dynamic *skip-breaks*
+  "Boolean or vector to determine whether to skip a breakpoint.
+  Don't set or examine this directly, it is bound in the session
+  binding map, use `skip-breaks!` and `skip-breaks?` instead.
+  Its value is discarded at the end each eval session."
+  nil)
+
+(defn- skip-breaks?
+  "True if the breakpoint at coordinates should be skipped.
+  If *skip-breaks* is true, return true.
+  If *skip-breaks* is a vector of integers, return true if coordinates
+  are deeper than this vector."
+  [coordinates]
+  (when-let [sb (@(:session *msg*) #'*skip-breaks*)]
+    (or
+     ;; From :continue, skip everything.
+     (true? sb)
+     ;; From :out, skip some breaks.
+     (let [parent (take (count sb) coordinates)]
+       (and (= sb parent)
+            (> (count coordinates) (count parent)))))))
+
+(defn- skip-breaks!
+  "Set the value of *skip-breaks* in the session binding map."
+  [bool-or-vec]
+  (swap! (:session *msg*) assoc #'*skip-breaks* bool-or-vec))
+
+(defn- abort!
+  "Stop current eval thread.
+  This does not quit the repl, it only interrupts an eval operation."
+  []
+  (.stop (:thread (meta (:session *msg*)))))
+
+;;;; ## Getting user input
 ;;; `wrap-debug` receives an initial message from the client, stores
-;;; it in `debugger-message`, and `breakpoint` answers it.
-(def debugger-message (atom nil))
+;;; it in `debugger-message`, and `breakpoint` answers it when asking
+;;; for input.
+(def debugger-message
+  "The message being used to communicate with the client.
+  Stored by the \"init-debugger\" op, and used by `read-debug` to ask
+  for debug input through the :need-debug-input status."
+  (atom nil))
 
-(defn wrap-debug [h]
-  (fn [{:keys [op force] :as msg}]
-    (if (= "init-debugger" op)
-      (let [stored-message @debugger-message]
-        (if (and stored-message (not force))
-          (transport/send (:transport msg)
-                          (response-for msg :status :done))
-          (do (when stored-message
-                (transport/send (:transport stored-message)
-                                (response-for stored-message :status :done)))
-              ;; TODO: Redefine this in the session binding map, see inspect.clj.
-              (reset! debugger-message msg))))
-      (h msg))))
+(def debugger-input
+  "Map atom holding all unprocessed debug inputs.
+  This is where the \"debug\" op stores replies received for debug
+  input requests. `read-debug` will halt until it finds its input in
+  this map (identified by a key), and will `dissoc` it afterwards."
+  (atom nil))
+
+(defn- read-debug
+  "Like `read`, but reply is sent through `debugger-message`.
+  type is sent in the message as :input-type."
+  [extras type]
+  (let [key (rand-int 10000000)]
```

Comment:
This looks like an id of some sort.  If that's the case how about using: `(str (java.util.UUID/randomUUID))` instead?